### PR TITLE
feat: add spot termination watcher (beta)

### PR DESCRIPTION
**Pull request by npalm**
This PR is adding a lambda function to watch termination events.

- Log instance information for termination warnings
- Create optional a metric with dimensions for the environment and instance type.

This PR limits to only checking the termination warning. Later we can extend on also start acting on terminations.

## Testing
Spot termination can be tested by initiate a termination event via the Spot Request overview (or cli).

## Todo

- [x] Write docs
- [x] Add to multi runner
- [ ] Describe next steps in an issue.

### CODE REVIEW

**Comment by npalm** [comment]
@GuptaNavdeep1983 @ScottGuymer @nadavsinai-philips @saragerion Do one of you have any clue why middy is let failing the unit tests. For modules like `lambdas/functions/ami-housekeeper` or `lamdas/functions/webhook`. The test is `lambda.test.ts` are working, where `lambda.te` also initiates middy.

In `lambdas/functions/spot-termination-watcher` the tests are failing with the error:

```
    TypeError: Cannot destructure property 'before' of 'middleware' as it is undefined.

      14 | import { Config } from './ConfigResolver';
      15 |
    > 16 | middy(interruptionWarning).use(captureLambdaHandler(tracer)).use(logMetrics(metrics));
         |                            ^
      17 | const config = new Config();
      18 |
      19 | export async function interruptionWarning(

      at Function.middy.use (../../node_modules/@middy/core/index.cjs:90:21)
      at Object.use (src/lambda.ts:16:28)
      at Object.<anonymous> (src/lambda.test.ts:6:1)
```

Even without adding the metrics part, which is the only difference comparing to the other modules, the tests are failing. CI is failing as well https://github.com/philips-labs/terraform-aws-github-runner/actions/runs/8129056369/job/22215682414

**Comment by npalm** [comment]
> @GuptaNavdeep1983 @ScottGuymer @nadavsinai-philips @saragerion Do one of you have any clue why middy is let failing the unit tests. For modules like `lamdas/functions/webhook`. The test is `lambda.test.ts` are working, where `lambda.te` also initiates middy.
> 
> In `lambdas/functions/spot-termination-watcher` the tests are failing with the error:
> 
> ```
>     TypeError: Cannot destructure property 'before' of 'middleware' as it is undefined.
> 
>       14 | import { Config } from './ConfigResolver';
>       15 |
>     > 16 | middy(interruptionWarning).use(captureLambdaHandler(tracer)).use(logMetrics(metrics));
>          |                            ^
>       17 | const config = new Config();
>       18 |
>       19 | export async function interruptionWarning(
> 
>       at Function.middy.use (../../node_modules/@middy/core/index.cjs:90:21)
>       at Object.use (src/lambda.ts:16:28)
>       at Object.<anonymous> (src/lambda.test.ts:6:1)
> ```
> 
> Even without adding the metrics part, which is the only difference comparing to the other modules, the tests are failing. CI is failing as well https://github.com/philips-labs/terraform-aws-github-runner/actions/runs/8129056369/job/22215682414

Fixed, figured out the problem. Tracer handler could be undefined in unit test (at least).

File: .github/workflows/update-docs.yml [hunk 1]
```diff
@@ -16,10 +16,32 @@ jobs:
     name: Auto update terraform docs
     runs-on: ubuntu-latest
     steps:
+      - uses: philips-software/app-token-action@9f5d57062c9f2beaffafaa9a34f66f824ead63a9 # v2.0.0
```
Comment (npalm) [inline .github/workflows/update-docs.yml:19]: Commit via a app instead of bot to ensure all workflows are executed on PR from maintainers

File: docs/examples/termination-watcher.md [hunk 2]
```diff
@@ -0,0 +1 @@
+--8<-- "examples/termination-watcher/README.md"
```
Comment (ScottGuymer) [inline docs/examples/termination-watcher.md:1]: is this file intended?

**Review by ScottGuymer — APPROVED** [review]
Its a bit of a long one.

tried to take a look at the important stuff and it all seems ok to me.

:shipit:

File: docs/examples/termination-watcher.md (hunk 2 above)
Comment (npalm) [reply docs/examples/termination-watcher.md:1]: yes but it should have some ascii ;)

File: docs/index.md [hunk 3]
```diff
@@ -64,6 +64,9 @@ The control plane (scale up lambda) will store the runner registration configura
 
 The AMI cleaner is a lambda that will clean up AMIs that are older than a configurable amount of days. This is useful when using the AMI builder to create AMIs. The cleaner will also check which AMIs are used the latest version of the launch template. And you can provide SSM config paths pointing to AMI IDs. The cleaner will not delete these AMIs. The AMI cleaner is opt in, it will not be created by default.
 
+### Spot Instance Termination Watcher
+
+The Spot Instance Termination Wachter is creating log and optional merrics for Spot Instance Termination warning sent by AWS two minutes before termination. The Lambda only will log instances details for instances tagged with `ghr:createdBy`. The module is by default not enabled.
```
Comment (GuptaNavdeep1983) [inline docs/index.md:69]: above section you said instances tagged with ghr:environment? Can we be consistent?

File: docs/configuration.md [hunk 4]
```diff
@@ -175,6 +175,21 @@ This tracing config generates timelines for following events:
 
 This feature has been disabled by default.
 
+## Termination watcher
+
+This feature is in early stage and therefore disabled by default.
+
+The termination watcher is currently watchting for spot termination notifications. The module is only taken events into account for instances tagged with `ghr:environment` by default when deployment the module as part of one of the main modules (root or multi-runner). The module can also be deployed stand-alone, in that case the tag filter needs to be tunned.
```
Comment (GuptaNavdeep1983) [inline docs/configuration.md:182]: below you used ghr:createdBy and here ghr:environment!

File: examples/termination-watcher/README.md [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+```
+
+Next switch to this example directory.
+
+```bash
+terraform init
+terraform apply
+```
+
+Once a Spot instance is terminated a log line and metric will be updated. Spot termiaation can be simulated by navigating to the request and invoce an interruption.
```
Comment (GuptaNavdeep1983) [inline examples/termination-watcher/README.md:23]: isnt it nice to detail out the simulation in details?

Comment (npalm) [reply examples/termination-watcher/README.md]: added some directions

File: docs/examples/termination-watcher.md (hunk 2 above)
Comment (npalm) [reply docs/examples/termination-watcher.md:1]: this file is including the example.

**Review by ScottGuymer — DISMISSED** [review]
